text/0066-separate-context-propagation.md

```diff
@@ -0,0 +1,553 @@
+
+
+# OpenTelemetry layered architecture
+
+The design of OpenTelemetry is based on the priciples of [aspect-oriented 
+programming](https://en.wikipedia.org/wiki/Aspect-oriented_programming), 
+adopted to the needs of distributed systems.
+
+OpenTelemetry is separated into two layers. The top layer contains a set of 
+independent **cross-cutting concerns**, which intertwine with a program's 
+application logic, and cannot be cleanly encapsulated. All concerns share an 
+underlying distributed **context propagation** layer, for storing state and 
+accessing data across the lifespan of a distributed transaction.
+
+
+# Cross-Cutting Concerns
+
+## Observability API
+Distributed tracing is one example of a cross-cutting concern. Tracing code is 
+interleaved with regular code, and ties together independent code modules which 
+would otherwise remain encapsulated. Tracing is also distributed, and requires 
+non-local, transaction-level context propagation in order to execute correctly.
+
+The various observability APIs are not described here directly. However, in this new 
+design, all observability APIs would be modified to make use of the generalized 
+context propagation mechanism described below, rather than the tracing-specific 
+propagation system it uses today.
+
+Note that OpenTelemetry APIs calls should *always* be given access to the entire 
+context object, and never just a subset of the context, such as the value in a 
+single key. This allows the SDK to make improvements and leverage additional 
+data that may be available, without changes to all of the call sites.
+
+## Correlations API
+
+In addition to trace propagation, OpenTelemetry provides a simple mechanism for 
+propagating arbitrary transaction-level data, called Correlations. Correlations 
+implements an extremely basic form of cross-cutting concern - a transaction-level 
+key-value store. Values placed in the 
+
+Event correlation is a mechanism for indexing events in one service with 
+attributes provided by a prior service in the same transaction. This helps to 
+establish a causal relationship between these events. For example, determining 
+that a particular browser version is associated with a failure in an image 
+processing service.
+
+Correlations is based on the [W3C Correlation-Context specification](https://w3c.github.io/correlation-context/), 
+and implements the protocol as it is defined in that working group.
+
+While Correlations can be used to prototype other cross-cutting concerns, this 
+mechanism is primarily intended to convey values for the OpenTelemetry 
+observability systems. New concerns with different criteria should be modeled 
+seperately, using the same underlying context propagation layer as building 
+blocks.
+
+For backwards compatibility, OpenTracing Baggage is propagated as Correlations 
+when using the OpenTracing bridge.
+
+As every correlation adds additional overhead to every request, correlations 
+should be used with care.
```

Comment:
what about delivery guarantees that implementation SHOULD/MUST provide?

Reply:
Added a MUST guarantee. Either the user or operator must explicitly choose which values to drop.